keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -288,9 +288,39 @@ like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey
   - xref [analysis of possible interactions with immutability and
     validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
 
-- If a object property collides with a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
-  it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision, all properties with a `_` prefix will always be
+- If a object property name is a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
+  it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision,  properties named with a CEL keyword and a  `_` prefix will be
   prefixed by `__` (generally, N+1 the existing number of `_`s).
+  
+- If a object property name is a CEL language identifier (`int`, `uint`, `double`, `bool`, `string`,
+  `bytes`, `list`, `map`, `null_type`, `type`, see [CEL language
+  identifiers](https://github.com/google/cel-spec/blob/master/doc/langdef.md#values)) it is not
+  accessible as a root variable and must be accessed via `self`, .e.g. `self.int`.
```

Comment:
do we really want to allow referencing things without `self.field` at all?  Has golang defeated me and made an explicit `this` seem normal in my world?

Reply:
This might be a bigger topic, but my current take:

Pros:
- It does a lot for keeping simple expressions short

Cons:
- we can't access this short list of keywords (I feel like this is a pretty minor thing)
- you can't do `has(field)` (but you can do `has(self.field)`), and more generally, there seem to be differences between the a "variable" and a "field" that bleed through (I'll explain that below)

Examples:

```
// in scope is an object with a field `a` that is unset:

has(a) -> "invalid argument to has() macro"
has(a.nestedField) -> "no such attribute: id: 2, names: [a]"
has(self.a) -> false
has(self.a.nestedField) -> "no such key: a"
has(self.a) && has(self.a.nestedField) -> false
```

The reason for the behavior differences is that `a` is a "bound variable" from CEL's perspective where-as `self.a` is a field.

I'll be honest that I really don't like the behavioral differences here, so I'd be willing restrict all field access to `self` is there is consensus.

(And I hate to ask, but is `self` still the best name for the variable if we only have a single variable? for all expressions).

cc @cici37 @liggitt @sttts

Reply:
+1 to apply the restriction of all field access vi `self` since 1)the behavior is consistent 2) users will not be confused by restriction of language identifiers 3) hide CEL details from users since users should not be expected to know `bound variable`/`field` differences.

Reply:
From offline discussions, there seems to be consensus that we should require all fields be accessed. I've update this PR and the implementation (https://github.com/kubernetes/kubernetes/pull/106051) to match.

Reply:
that seems reasonable. it also has the side benefit of keeping the top-level identifier namespace free for future additions like `oldSelf` without breaking previously accepted expressions